core/CronArchive/SegmentArchiving.php

```diff
@@ -0,0 +1,270 @@
+<?php
+/**
+ * Piwik - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ */
+namespace Piwik\CronArchive;
+
+use Doctrine\Common\Cache\Cache;
+use Matomo\Cache\Transient;
+use Piwik\Common;
+use Piwik\Container\StaticContainer;
+use Piwik\CronArchive;
+use Piwik\Date;
+use Piwik\Db;
+use Piwik\Period\Range;
+use Piwik\Plugins\SegmentEditor\Model;
+use Piwik\Segment;
+use Piwik\Site;
+use Psr\Log\LoggerInterface;
+
+/**
+ * Provides URLs that initiate archiving during cron archiving for segments.
+ *
+ * Handles the `[General] process_new_segments_from` INI option.
+ */
+class SegmentArchiving
+{
+    const BEGINNING_OF_TIME = 'beginning_of_time';
+    const CREATION_TIME = 'segment_creation_time';
+    const LAST_EDIT_TIME = 'segment_last_edit_time';
+    const DEFAULT_BEGINNIN_OF_TIME_LAST_N_YEARS = 7;
+
+    /**
+     * @var Model
+     */
+    private $segmentEditorModel;
+
+    /**
+     * @var Transient
+     */
+    private $segmentListCache;
+
+    /**
+     * @var Date
+     */
+    private $now;
+
+    private $processNewSegmentsFrom;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var int
+     */
+    private $beginningOfTimeLastNInYears;
+
+    public function __construct($processNewSegmentsFrom, $beginningOfTimeLastNInYears = self::DEFAULT_BEGINNIN_OF_TIME_LAST_N_YEARS,
+                                Model $segmentEditorModel = null, Cache $segmentListCache = null, Date $now = null,
+                                LoggerInterface $logger = null)
+    {
+        $this->processNewSegmentsFrom = $processNewSegmentsFrom;
+        $this->beginningOfTimeLastNInYears = $beginningOfTimeLastNInYears;
+        $this->segmentEditorModel = $segmentEditorModel ?: new Model();
+        $this->segmentListCache = $segmentListCache ?: new Transient();
+        $this->now = $now ?: Date::factory('now');
+        $this->logger = $logger ?: StaticContainer::get('Psr\Log\LoggerInterface');
+    }
+
+    public function getSegmentArchivesToInvalidateForNewSegments($idSite)
+    {
+        $result = [];
+
+        $segmentsForSite = $this->getAllSegments();
+        foreach ($segmentsForSite as $storedSegment) {
+            if (empty($storedSegment['auto_archive'])) {
```

Comment:
BTW... minor tweak that is missing in 3.x-dev...

if browser archiving is disabled, and browser segment archiving as well (`browser_archiving_disabled_enforce=1`) then we should assume all segments are supposed to be auto archived cause otherwise a segment would never be automatically archived. This is basically only an issue if you have segments for browser archiving, and then change the ini setting `browser_archiving_disabled_enforce` from `0` to `1`. This avoids that the user manually needs to change the database flag for these segments.